Exception transparency in handler/callback interfaces

qualidafial opened the pull request:
Addresses issue #290 

To do:

* [x] ~~Remove `Transaction` callback interface in favor of `TransactionCallback`~~ Rename it to `TransactionalCallback` and move it to sqlobject where it belongs
* [x] Add `TransactionalConsumer` and friend `Transactional.useTransaction(TransactionalConsumer)`
* ~~See if `TransactionFailedException` can be removed~~ nah

stevenschlansker commented on core/src/main/java/org/jdbi/v3/tweak/transactions/LocalTransactionHandler.java:
```diff
@@ -165,36 +165,28 @@ public boolean isInTransaction(Handle handle)
-            throw new TransactionFailedException("Transaction failed do to exception being thrown " +
-                                                 "from within the callback. See cause " +
-                                                 "for the original exception.", e);
+            throw (X) e;
```
Ah, what a sneaky throw ;)

qualidafial replied:
My reasoning is that the try can only throw a `RuntimeException` or `X`. So the cast is safe from a runtime perspective. At compile time `X` is erased to just `Exception`.

Plus there's the whole part about how checked exceptions are a purely compile time construct--the runtime makes no distinction and any method can legally throw any Throwable they like.

stevenschlansker replied:
Yeah, got it.  I like that it removes one of the catch clauses!

stevenschlansker commented:
Nice!  Is this ready to merge, or still pending the TODOs?  I think the current state is good, happy to merge as is and handle the rest as followups, or block it -- your call.  Thanks!

qualidafial commented:
I want to hold this one and merge after the extensions, since the merge effort (and pain) should be less on this PR than on the other one.

qualidafial commented:
@stevenschlansker I made one more change: renamed `Transaction` interface to `TransactionalCallback`, moved it from core to sqlobject, fixed its generic signature. Also added `TransactionalConsumer` interface to go with it.

If you're good with these changes, I think this is ready to merge.
